### ISSUE TITLE
'this' banned term conflicts with Typescript's function this-types

### ISSUE THREAD

**Issue by dsole** [Microsoft/tslint-microsoft-contrib#261]
I'm finding that the function level this-type introduced by [Microsoft/TypeScript #6739](https://github.com/Microsoft/TypeScript/pull/6739) is triggering the "noReservedKeywords" rule.

Using this example from the PR
```
interface I {
  n: number;
  method: (m: number) => number;
}
function futureMethod(this: I, m: number) {
  return this.n + m;
}
let o: I = { n: 12, method: futureMethod };
```

The this-type feature of TS declares the futureMethod has having one parameter (`m`) and that the 
"`this`" object within its scope implements interface `I`.
But instead it has a tslint error:
> [tslint] Forbidden reference to reserved keyword: this
>(parameter) this: I

It feels like a special exception is needed for the this keyword, to allow for function this-types.

**Comment by HamletDRC** [comment]
I see that this rule will interface with the Polymorphic this Type described here: http://www.typescriptlang.org/docs/handbook/advanced-types.html

I don't see any official documentation on the feature you mention. Is it a TypeScript 2.0 feature?

**Comment by dsole** [comment]
Now that TS 2.0 is out, I can confidently say that yes it is going to be included.
They're mentioned in the [release notes](http://www.typescriptlang.org/docs/release-notes/typescript-2.0.html) under "[Specifying the type of this for functions](http://www.typescriptlang.org/docs/release-notes/typescript-2.0.html#specifying-the-type-of-this-for-functions)"
